File: linkedlist/Q61.java
public class Q61 {
    public ListNode rotateRight(ListNode head, int k) {
        if (head == null || head.next == null || k == 0) return head;

        ListNode ptr = head;
        int length = 1;
        while (ptr.next != null) {
            ptr = ptr.next;
            length++;
        }
        ListNode tail = ptr;

        if (k % length == 0) return head;

        int count = length - k % length - 1;
        ptr = head;
        while (count-- > 0) {
            ptr = ptr.next;
        }
        ListNode newHead = ptr.next;
        ptr.next = null;
        tail.next = head;
        return newHead;
    }
}

File: LinkedList-Stack-Queue/Q61.py
# https://leetcode.com/problems/rotate-list/

# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def rotateRight(self, head: ListNode, k: int) -> ListNode:
        if not head or not head.next:
            return head

        length = 1
        tail = head
        while tail.next:
            length += 1
            tail = tail.next

        k %= length
        if k == 0:
            return head

        leftTail = head
        for _ in range(length - k - 1):
            leftTail = leftTail.next

        rightHead = leftTail.next
        leftTail.next = None
        tail.next = head
        return rightHead
